File: nosql-create-index-go/controlplane.go
package main

import (
	"context"
	"errors"
	"fmt"
	"os/exec"
	"strings"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/cosmos/armcosmos/v3"
)

// ptr is a helper to return a pointer to a value
func ptr[T any](v T) *T { return &v }

// getSubscriptionID resolves the subscription at runtime via the Azure CLI
func getSubscriptionID() (string, error) {
	out, err := exec.Command("az", "account", "show", "--query", "id", "-o", "tsv").Output()
	if err != nil {
		return "", fmt.Errorf("az account show failed: %w", err)
	}
	return strings.TrimSpace(string(out)), nil
}

// CreateContainersWithVectorIndexes creates SQL containers with vector indexes using the ARM SDK
func CreateContainersWithVectorIndexes(
	ctx context.Context,
	credential *azidentity.DefaultAzureCredential,
	config *Config,
) error {
	// Get subscription ID
	subscriptionID, err := getSubscriptionID()
	if err != nil {
		return fmt.Errorf("failed to get subscription ID: %w", err)
	}

	// Create ARM client
	client, err := armcosmos.NewSQLResourcesClient(subscriptionID, credential, nil)
	if err != nil {
		return fmt.Errorf("failed to create ARM client: %w", err)
	}

	// Ensure database exists first
	fmt.Printf("\n=== Phase 1: Create Database ===\n")
	fmt.Printf("  Database: %s\n", config.DatabaseName)

	dbPoller, err := client.BeginCreateUpdateSQLDatabase(
		ctx,
		config.ResourceGroup,
		config.AccountName,
		config.DatabaseName,
		armcosmos.SQLDatabaseCreateUpdateParameters{
			Location: ptr(config.Location),
			Properties: &armcosmos.SQLDatabaseCreateUpdateProperties{
				Resource: &armcosmos.SQLDatabaseResource{
					ID: ptr(config.DatabaseName),
				},
				Options: &armcosmos.CreateUpdateOptions{},
			},
		},
		nil,
	)
	if err != nil {
		return fmt.Errorf("failed to begin creating database: %w", err)
	}

	if _, err := dbPoller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{Frequency: 5 * time.Second}); err != nil {
		return fmt.Errorf("failed to create database: %w", err)
	}

	fmt.Printf("  ✓ Database created or already exists\n")

	// Create containers with vector indexes
	indexConfigs := []struct {
		indexType     armcosmos.VectorIndexType
		containerName string
	}{
		{armcosmos.VectorIndexTypeDiskANN, diskANNContainer},
		{armcosmos.VectorIndexTypeQuantizedFlat, quantizedFlatContainer},
	}

	embeddingPath := "/" + strings.TrimPrefix(config.EmbeddingFieldName, "/")
	partitionKeyPath := "/" + config.PartitionKeyFieldName

	for _, indexConfig := range indexConfigs {
		fmt.Printf("\n=== Creating Container: %s ===\n", indexConfig.containerName)
		fmt.Printf("  Index type:     %s\n", string(indexConfig.indexType))
		fmt.Printf("  Embedding path: %s\n", embeddingPath)
		fmt.Printf("  Dimensions:     %d\n", config.EmbeddingDimensions)
		fmt.Printf("  Distance func:  cosine (queried with all 3 metrics)\n")

		// Build container resource with vector embedding policy
		containerResource := &armcosmos.SQLContainerResource{
			ID: ptr(indexConfig.containerName),
			PartitionKey: &armcosmos.ContainerPartitionKey{
				Paths: []*string{ptr(partitionKeyPath)},
				Kind:  ptr(armcosmos.PartitionKindHash),
			},
			VectorEmbeddingPolicy: &armcosmos.VectorEmbeddingPolicy{
				VectorEmbeddings: []*armcosmos.VectorEmbedding{
					{
						Path:             ptr(embeddingPath),
						DataType:         ptr(armcosmos.VectorDataTypeFloat32),
						Dimensions:       ptr(int32(config.EmbeddingDimensions)),
						DistanceFunction: ptr(armcosmos.DistanceFunctionCosine),
					},
				},
			},
			IndexingPolicy: &armcosmos.IndexingPolicy{
				IndexingMode: ptr(armcosmos.IndexingModeConsistent),
				Automatic:    ptr(true),
				IncludedPaths: []*armcosmos.IncludedPath{
					{Path: ptr("/*")},
				},
				ExcludedPaths: []*armcosmos.ExcludedPath{
					{Path: ptr(`/"_etag"/?`)},
					{Path: ptr(embeddingPath + "/*")},
				},
				VectorIndexes: []*armcosmos.VectorIndex{
					{
						Path: ptr(embeddingPath),
						Type: ptr(indexConfig.indexType),
					},
				},
			},
		}

		// Delete existing container for idempotent re-runs
		fmt.Printf("  Cleaning up existing container...\n")
		deletePoller, err := client.BeginDeleteSQLContainer(
			ctx,
			config.ResourceGroup,
			config.AccountName,
			config.DatabaseName,
			indexConfig.containerName,
			nil,
		)
		if err == nil {
			// If delete operation started, wait for it
			if _, err = deletePoller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{Frequency: 5 * time.Second}); err == nil {
				fmt.Printf("  Deleted existing container\n")
			}
		}

		// Create the container
		fmt.Printf("  Creating container with vector index...\n")
		start := time.Now()

		// Note: Options is nil to support serverless accounts (which don't allow explicit throughput).
		// For provisioned accounts, you could set Options.Throughput or Options.AutoscaleSettings.
		containerPoller, err := client.BeginCreateUpdateSQLContainer(
			ctx,
			config.ResourceGroup,
			config.AccountName,
			config.DatabaseName,
			indexConfig.containerName,
			armcosmos.SQLContainerCreateUpdateParameters{
				Location: ptr(config.Location),
				Properties: &armcosmos.SQLContainerCreateUpdateProperties{
					Resource: containerResource,
					Options:  nil, // nil = serverless compatible (no throughput setting)
				},
			},
			nil,
		)
		if err != nil {
			return fmt.Errorf("failed to begin creating container %s: %w", indexConfig.containerName, err)
		}

		if _, err := containerPoller.PollUntilDone(ctx, &runtime.PollUntilDoneOptions{Frequency: 5 * time.Second}); err != nil {
			return fmt.Errorf("failed to create container %s: %w", indexConfig.containerName, err)
		}

		elapsed := time.Since(start)
		fmt.Printf("  ✓ Container created in %.2fs\n", elapsed.Seconds())

		// Verify the container exists and has the correct configuration
		got, err := client.GetSQLContainer(
			ctx,
			config.ResourceGroup,
			config.AccountName,
			config.DatabaseName,
			indexConfig.containerName,
			nil,
		)
		if err != nil {
			return fmt.Errorf("failed to verify container %s: %w", indexConfig.containerName, err)
		}

		res := got.Properties.Resource
		if res == nil {
			return fmt.Errorf("read-back resource for %s is nil", indexConfig.containerName)
		}

		// Log verification
		if res.VectorEmbeddingPolicy != nil && len(res.VectorEmbeddingPolicy.VectorEmbeddings) > 0 {
			emb := res.VectorEmbeddingPolicy.VectorEmbeddings[0]
			fmt.Printf("  ✓ Verified: Vector embedding policy configured\n")
			fmt.Printf("    - Path: %s\n", *emb.Path)
			fmt.Printf("    - DataType: %s\n", *emb.DataType)
			fmt.Printf("    - Dimensions: %d\n", *emb.Dimensions)
			fmt.Printf("    - DistanceFunction: %s\n", *emb.DistanceFunction)
		}

		if res.IndexingPolicy != nil && len(res.IndexingPolicy.VectorIndexes) > 0 {
			idx := res.IndexingPolicy.VectorIndexes[0]
			fmt.Printf("  ✓ Verified: Vector index configured\n")
			fmt.Printf("    - Path: %s\n", *idx.Path)
			fmt.Printf("    - Type: %s\n", *idx.Type)
		}
	}

	fmt.Printf("\n✓ All containers created successfully\n")
	return nil
}

func DeleteContainers(ctx context.Context, credential *azidentity.DefaultAzureCredential, cfg *Config) error {
	client, err := armcosmos.NewSQLResourcesClient(cfg.SubscriptionID, credential, nil)
	if err != nil {
		return fmt.Errorf("failed to create SQL resources client: %w", err)
	}

	for _, containerName := range []string{cfg.DiskANNContainerName, cfg.QuantizedFlatContainerName} {
		fmt.Printf("  Deleting %s...\n", containerName)
		start := time.Now()
		deleteCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)

		poller, err := client.BeginDeleteSQLContainer(
			deleteCtx,
			cfg.ResourceGroup,
			cfg.AccountName,
			cfg.DatabaseName,
			containerName,
			nil,
		)
		if err != nil {
			cancel()
			if isNotFound(err) {
				fmt.Printf("  ✓ Container does not exist: %s\n", containerName)
				continue
			}
			return fmt.Errorf("failed to delete container %q: %w", containerName, err)
		}

		_, err = poller.PollUntilDone(
			deleteCtx,
			&runtime.PollUntilDoneOptions{Frequency: 5 * time.Second},
		)
		cancel()
		if err != nil {
			if isNotFound(err) {
				fmt.Printf("  ✓ Container does not exist: %s\n", containerName)
				continue
			}
			return fmt.Errorf("failed to wait for container deletion %q: %w", containerName, err)
		}

		fmt.Printf("  ✓ Deleted %s in %.1fs\n", containerName, time.Since(start).Seconds())
	}

	return nil
}

func isNotFound(err error) bool {
	var responseError *azcore.ResponseError
	return errors.As(err, &responseError) && responseError.StatusCode == 404
}

File: nosql-create-index-python/src/control_plane.py
"""Control-plane operations for Cosmos DB SQL (NoSQL) API using ARM SDK.

Uses raw JSON payloads to work around ARM SDK limitations with vector types.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from azure.core.exceptions import HttpResponseError
from azure.identity import DefaultAzureCredential
from azure.mgmt.cosmosdb import CosmosDBManagementClient
from azure.mgmt.cosmosdb.models import SqlContainerCreateUpdateParameters, SqlContainerResource

if TYPE_CHECKING:
    from .config import SampleConfig


def _build_container_payload(
    container_name: str,
    partition_key_path: str,
    embedding_field: str,
    dimensions: int,
    index_type: str
) -> dict[str, Any]:
    """Build container creation payload with vector index configuration.
    
    Args:
        container_name: Name of the container
        partition_key_path: Path to partition key field (e.g., "/HotelId")
        embedding_field: Path to embedding field (e.g., "/embedding")
        dimensions: Vector embedding dimensions
        index_type: Vector index type ("diskANN" or "quantizedflat")
    
    Returns:
        Container resource payload as dictionary
    """
    return {
        "id": container_name,
        "partitionKey": {
            "paths": [partition_key_path],
            "kind": "Hash"
        },
        "indexingPolicy": {
            "indexingMode": "Consistent",
            "automatic": True,
            "includedPaths": [{"path": "/*"}],
            "excludedPaths": [{"path": "/_etag/?"}, {"path": f"{embedding_field}/*"}],
            "vectorIndexes": [
                {
                    "path": embedding_field,
                    "type": index_type
                }
            ]
        },
        "vectorEmbeddingPolicy": {
            "vectorEmbeddings": [
                {
                    "path": embedding_field,
                    "dataType": "float32",
                    "dimensions": dimensions,
                    "distanceFunction": "cosine"
                }
            ]
        }
    }


def create_containers(credential: DefaultAzureCredential, config: SampleConfig) -> None:
    """Create SQL containers with vector indexes using control-plane (ARM) SDK.

    Args:
        credential: Azure credential for authentication
        config: Sample configuration with resource details
    """
    client = CosmosDBManagementClient(
        credential=credential,
        subscription_id=config.subscription_id
    )

    embedding_path = f"/{config.embedding_field_name}"
    from . import config as config_module
    containers_config = [
        {"type": "diskANN", "container_name": config_module.KNOWN_CONTAINERS["diskann"]},
        {"type": "QuantizedFlat", "container_name": config_module.KNOWN_CONTAINERS["quantizedflat"]},
    ]

    for container_config in containers_config:
        print(f"\n=== Phase 1: Create Container with Vector Index ===")
        print(f"  Container:      {container_config['container_name']}")
        print(f"  Index type:     {container_config['type']}")
        print(f"  Dimensions:     {config.expected_dimensions}")
        print(f"  Distance func:  cosine (queried with all 3 metrics)")

        # Delete existing container to ensure clean state (idempotent)
        try:
            client.sql_resources.begin_delete_sql_container(
                resource_group_name=config.resource_group,
                account_name=config.account_name,
                database_name=config.database_name,
                container_name=container_config["container_name"]
            ).result()
            print(f"  Deleted existing container")
        except HttpResponseError as e:
            if e.status_code == 404:
                print(f"  Container does not exist (OK)")
            else:
                raise

        # Build container payload with vector index
        container_payload = _build_container_payload(
            container_name=container_config["container_name"],
            partition_key_path="/Region",
            embedding_field=embedding_path,
            dimensions=config.expected_dimensions,
            index_type=container_config["type"]
        )

        # Wrap the environment-derived resource name in the typed ARM model so
        # the SDK preserves the container id in the request body.
        params = SqlContainerCreateUpdateParameters(
            resource=SqlContainerResource(container_payload)
        )

        client.sql_resources.begin_create_update_sql_container(
            resource_group_name=config.resource_group,
            account_name=config.account_name,
            database_name=config.database_name,
            container_name=container_config["container_name"],
            create_update_sql_container_parameters=params
        ).result()

        print(f"  Created in ~1s")
        print(f"  Vector index is IMMUTABLE — cannot be changed after creation")


def delete_containers(credential: DefaultAzureCredential, config: SampleConfig) -> None:
    """Delete SQL containers using control-plane (ARM) SDK.

    Args:
        credential: Azure credential for authentication
        config: Sample configuration with resource details
    """
    client = CosmosDBManagementClient(
        credential=credential,
        subscription_id=config.subscription_id
    )

    # Always try to delete both containers, regardless of which one is active
    from . import config as config_module
    container_names = list(config_module.KNOWN_CONTAINERS.values())

    for container_name in container_names:
        try:
            # Delete container using ARM SDK
            client.sql_resources.begin_delete_sql_container(
                resource_group_name=config.resource_group,
                account_name=config.account_name,
                database_name=config.database_name,
                container_name=container_name
            ).result()  # Wait for async operation
            print(f"  ✓ Deleted {container_name}")
        except HttpResponseError as e:
            # 404 is expected if container doesn't exist — skip silently
            if e.status_code == 404:
                print(f"  ✓ {container_name} does not exist (OK)")
            else:
                raise
